cli/commands/scaffold/action.go

```diff
@@ -221,62 +221,93 @@ func Run(ctx context.Context, opts *options.TerragruntOptions, moduleURL, templa
+	}
 
-	if templateURL != "" {
-		// process template url if was passed
-		parsedTemplateURL, err := tf.ToSourceURL(templateURL, tempDir)
-		if err != nil {
-			return "", errors.New(err)
-		}
+	if err := os.WriteFile(util.JoinPath(boilerplateDir, "boilerplate.yml"), []byte(DefaultBoilerplateConfig), ownerWriteGlobalReadPerms); err != nil {
+		return "", errors.New(err)
+	}
 
-		parsedTemplateURL, err = rewriteTemplateURL(ctx, opts, parsedTemplateURL)
-		if err != nil {
-			return "", errors.New(err)
-		}
-		// regenerate template url with all changes
-		templateURL = parsedTemplateURL.String()
+	return boilerplateDir, nil
+}
 
-		// prepare temporary directory for template
-		templateDir, err = os.MkdirTemp("", "template")
-		if err != nil {
-			return "", errors.New(err)
-		}
+// parseTemplateURL - parse URL and download files
+func parseTemplateURL(ctx context.Context, opts *options.TerragruntOptions, templateURL string, tempDir string) (string, error) {
+	parsedTemplateURL, err := tf.ToSourceURL(templateURL, tempDir)
+	if err != nil {
+		return "", errors.New(err)
+	}
 
-		// downloading template
-		opts.Logger.Infof("Using template from %s", templateURL)
+	parsedTemplateURL, err = rewriteTemplateURL(ctx, opts, parsedTemplateURL)
+	if err != nil {
+		return "", errors.New(err)
+	}
+	// regenerate template url with all changes
+	templateURL = parsedTemplateURL.String()
 
-		if _, err := getter.GetAny(ctx, templateDir, templateURL); err != nil {
-			return "", errors.New(err)
-		}
+	// prepare temporary directory for template
+	templateDir, err := os.MkdirTemp("", "template")
+	if err != nil {
+		return "", errors.New(err)
+	}
+
+	// downloading templateURL
+	opts.Logger.Infof("Using template from %s", templateURL)
+
+	if _, err := getter.GetAny(ctx, templateDir, templateURL); err != nil {
+		return "", errors.New(err)
 	}
-	// prepare boilerplate dir
+
+	return templateDir, nil
+}
+
+// prepareBoilerplateFiles - prepare boilerplate files from provided template, tf module, or (custom) default template
+func prepareBoilerplateFiles(ctx context.Context, opts *options.TerragruntOptions, templateURL string, tempDir string) (string, error) {
 	boilerplateDir := util.JoinPath(tempDir, util.DefaultBoilerplateDir)
-	// use template dir as boilerplate dir
-	if templateDir != "" {
-		boilerplateDir = templateDir
+
+	// process template url if it was passed. This overrides the .boilerplate folder in the OpenTofu/Terraform module
+	if templateURL != "" {
+		// process template url if it was passed
+		tempTemplateDir, err := parseTemplateURL(ctx, opts, templateURL, tempDir)
+		if err != nil {
+			return "", errors.New(err)
+		}
+
+		boilerplateDir = tempTemplateDir
 	}
 
 	// if boilerplate dir is not found, create one with default template
 	if !files.IsExistingDir(boilerplateDir) {
-		// no default boilerplate dir, create one
-		defaultTempDir, err := os.MkdirTemp("", "boilerplate")
+		config, err := config.ReadCatalogConfig(ctx, opts)
```

Comment:
Is this the right behavior here? Should scaffold always fail if the catalog config discovered in the parent directory is invalid?

Reply:
That's a good question. I'll leave that to your judgment. If we get to this point in the code, no Boilerplate template was provided as a cli input, and the terraform module also doesn't contain a `.boilerplate` folder. That leaves one of two options: either we use the custom default template in the catalog config (if there is one), or we stick to the hardcoded default template.

I can see the argument for falling back to the hardcoded custom template in case of an invalid catalog config, but that may surprise a user who thought they defined a custom default config. I would personally lean towards flagging the issue with the catalog config though, even if it doesn't actually contain a custom template. As a user I'd like to know if my config is invalid. Sooner rather than later.